File: examples/ts/src/shared/remotes.ts
import { Client, Server, createRemotes, loggerMiddleware, remote } from "@rbxts/remo";
import { t } from "@rbxts/t";

const todo = t.string;
const todoList = t.array(todo);

export const remotes = createRemotes(
	{
		// add a todo to the server
		addTodo: remote<Server, [name: string]>(todo),

		// remove a todo from the server
		removeTodo: remote<Server, [name: string]>(todo),

		// get all todos from the server
		getTodos: remote<Server, []>().returns<string[]>(todoList),

		// send new todos to clients
		todosChanged: remote<Client, [todos: string[]]>(),
	},
	loggerMiddleware,
);

File: examples/lua/src/shared/remotes.lua
local ReplicatedStorage = game:GetService("ReplicatedStorage")

local Remo = require(ReplicatedStorage.Packages.Remo)
local t = require(ReplicatedStorage.Packages.t)

type Remotes = Remo.Remotes<{
	addTodo: Remo.ClientToServer<string>,
	removeTodo: Remo.ClientToServer<string>,
	getTodos: Remo.ServerAsync<(), ({ string })>,
	todosChanged: Remo.ServerToClient<{ string }>,
}>

local todo = t.string
local todoList = t.array(todo :: any)

local remotes: Remotes = Remo.createRemotes({
	addTodo = Remo.remote(todo),
	removeTodo = Remo.remote(todo),
	getTodos = Remo.remote().returns(todoList),
	todosChanged = Remo.remote(todoList),
}, Remo.loggerMiddleware)

return remotes
